core/state_processor.go

```diff
@@ -170,14 +194,22 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 						return nil, nil, 0, fmt.Errorf("error decoding bytes %x: %w", stIt.Slot(), err)
 					}
 					copy(safeValue[32-len(value):], value)
-					slotnr := rawdb.ReadPreimage(statedb.Database().DiskDB(), stIt.Hash())
 
-					mkv.addStorageSlot(addr, slotnr, safeValue[:])
+					var slotnr [32]byte
+					if _, err := io.ReadFull(fpreimages, slotnr[:]); err != nil {
+						return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+					}
+					if crypto.Keccak256Hash(slotnr[:]) != stIt.Hash() {
+						return nil, nil, 0, fmt.Errorf("preimage file does not match storage hash: %s!=%s", crypto.Keccak256Hash(slotnr[:]), stIt.Hash())
+					}
+					preimageSeek += int64(len(slotnr))
```

Comment:
We keep in `preimageSeek` the current pointer in the file. When we finish with the block, we'll see how we use this variable to call `SetCurrentPreimageOffset(...)`.